### PR TITLE
Implement double backwards for a number of Index/Masked/Sort functions

### PR DESCRIPTION
**Pull request by gchanan**
Full list of functions: IndexAdd, IndexFill, IndexSelect, MaskedFill, MaskedSelect, Scatter, Sort, Topk.

### CODE REVIEW

**Comment by gchanan** [comment]
are these distributed test failures now?

**Comment by soumith** [comment]
yea something's weird with the distributed failures, i'll look, but they are known

File: torch/autograd/_functions/tensor.py [hunk 1]
```diff
@@ -191,7 +190,7 @@ def backward(ctx, grad_output):
 
         if ctx.needs_input_grad[3]:
             index, = ctx.saved_tensors
-            grad_tensor2 = grad_output.index_select(ctx.dim, index)
+            grad_tensor2 = grad_output.index_select(ctx.dim, Variable(index))
```
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: You should use `saved_variables` instead of repacking index (same for All other index functions)

File: torch/autograd/_functions/tensor.py [hunk 2]
```diff
@@ -434,13 +431,12 @@ def forward(ctx, tensor, mask):
...
 
     @staticmethod
-    @once_differentiable
     def backward(ctx, grad_output):
         mask, = ctx.saved_tensors
         grad_tensor = None
         if ctx.needs_input_grad[0]:
-            grad_tensor = grad_output.new(ctx.input_size).zero_()
-            grad_tensor.masked_scatter_(mask, grad_output)
+            grad_tensor = Variable(grad_output.data.new(ctx.input_size).zero_(), requires_grad=True)
```
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: This shouldn't require grad

File: torch/autograd/_functions/tensor.py [hunk 3]
```diff
@@ -415,12 +413,11 @@ def forward(ctx, tensor, mask, value, inplace=False):
         return tensor.masked_fill_(mask, value)
 
     @staticmethod
-    @once_differentiable
     def backward(ctx, grad_output):
         mask, = ctx.saved_tensors
         grad_tensor = None
         if ctx.needs_input_grad[0]:
-            grad_tensor = maybe_unexpand(grad_output.clone().masked_fill_(mask, 0), ctx.tensor_size)
+            grad_tensor = maybe_unexpand(grad_output.clone().masked_fill_(Variable(mask), 0), ctx.tensor_size)
```
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: Don't re-wrap mask

**Review by apaszke — COMMENTED** [review]
There are 3 issues with this PR. They all appear multiple times (every line should be changed), but I only commented once for each
